data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/s3/SecurityLakeBucketSelector.java

```diff
@@ -0,0 +1,85 @@
+public class SecurityLakeBucketSelector implements S3BucketSelector {
+    final private SecurityLakeBucketSelectorConfig securityLakeBucketSelectorConfig;
+
+    private S3SinkConfig s3SinkConfig;
+
+    private String pathPrefix;
+
+    private String sourceLocation;
+
+    @DataPrepperPluginConstructor
+    public SecurityLakeBucketSelector(final SecurityLakeBucketSelectorConfig securityLakeBucketSelectorConfig) {
+        this.securityLakeBucketSelectorConfig = securityLakeBucketSelectorConfig;
+    }
+
+    public void initialize(S3SinkConfig s3SinkConfig) {
+        this.s3SinkConfig = s3SinkConfig;
+        SecurityLakeClient securityLakeClient = SecurityLakeClient.create();
+        String arn = s3SinkConfig.getAwsAuthenticationOptions().getAwsStsRoleArn();
+        String principal = arn.split(":")[4];
+        String sourceName = securityLakeBucketSelectorConfig.getSourceName() != null ? securityLakeBucketSelectorConfig.getSourceName() : RandomStringUtils.randomAlphabetic(7);
+        CreateCustomLogSourceResponse response =
+                securityLakeClient.createCustomLogSource(
+                        CreateCustomLogSourceRequest.builder()
+                                .sourceName(sourceName+"_KK_"+RandomStringUtils.randomAlphabetic(4))
+                                .eventClasses(List.of(securityLakeBucketSelectorConfig.getLogClass()))
+                                .sourceVersion(securityLakeBucketSelectorConfig.getSourceVersion())
+                                .configuration(CustomLogSourceConfiguration.builder()
+                                        .crawlerConfiguration(CustomLogSourceCrawlerConfiguration.builder()
+                                                .roleArn(arn)
+                                                .build())
+                                        .providerIdentity(AwsIdentity.builder()
+                                                .externalId(securityLakeBucketSelectorConfig.getExternalId())
+                                                .principal(principal)
+                                                .build())
+                                        .build())
+                                .build());
+        CustomLogSourceProvider provider = response.source().provider();
+        this.sourceLocation = provider.location();
+        final String region=s3SinkConfig.getAwsAuthenticationOptions().getAwsRegion().toString();
+        final String accountId=arn.split(":")[4];
+        final LocalDate now = LocalDate.now();
+        final String eventDay = String.format("%d%02d%02d", now.getYear(), now.getMonthValue(), now.getDayOfMonth());
+        int locIdx = sourceLocation.indexOf("/ext/");
+        pathPrefix = sourceLocation.substring(locIdx+1)+"region="+region+"/accountId="+accountId+"/eventDay="+eventDay+"/";
+    }
+
+    public String getPathPrefix() {
+        return pathPrefix;
+    }
+
+    @Override
+    public Map<String, String> getMetadata(int eventCount) {
```

Comment:
I don't think this is related to the `bucket_selector`. I'd suggest that the `aws_security_lake` transformation just add this as additional metadata. I also think "additional metadata" is clearer since it doesn't overwrite the existing metadata.

```
s3:
  bucket_selector:
    aws_security_lake
  additional_metadata:
    asl_rows: ${numberOfEvents}
```

Reply:
@dlvenable how do we plan do support `${numberOfEvents}`? It is not a field in the key. It is not a function. I think it would be confusing

Reply:
I thought we had done something similar to this in either the `s3` sink or `opensearch` sink already where you could add some custom sink-only data (either to the index name or the S3 path). If we have that, we can follow that pattern.

Another idea is to make it somewhat less flexible by specifying the name exactly.

```
s3:
  additional_batch_metadata:
    aws_rows: numberOfEvents
```

We could require that the value be `numberOfEvents` exactly. In the future, maybe we add more.

Yet another approach would be less flexible, but could solve it:

```
s3:
  number_of_events_key: asl_rows
```